src/ulong_extras/test/t-mulmod_shoup.c

```diff
@@ -15,30 +16,29 @@
 
 TEST_FUNCTION_START(n_mulmod_shoup, state)
 {
-    int i, result;
-
-    for (i = 0; i < 10000 * flint_test_multiplier(); i++)
+    for (ulong i = 0; i < 10000 * flint_test_multiplier(); i++)
     {
-        ulong a, b, d, r1, r2, q, p1, p2, w_pr;
-
-        d = n_randtest_not_zero(state) / 2 + 1;
-        a = n_randtest(state) % d;
-        b = n_randtest(state) % d;
 
-        w_pr = n_mulmod_precomp_shoup(a, d);
+        const ulong d = n_randtest_not_zero(state) / 2 + 1;  // 0 < d < 2**(FLINT_BITS-1)
+        const ulong a = n_randtest(state) % d;  // a must be < d
+        const ulong b = n_randtest(state);  // b is arbitrary
 
-        r1 = n_mulmod_shoup(a, b, w_pr, d);
+        // mulmod_shoup
+        const ulong a_pr = n_mulmod_precomp_shoup(a, d);
+        const ulong r1 = n_mulmod_shoup(a, b, a_pr, d);
 
+        // trivial mulmod
+        ulong r2, q, p1, p2;
         umul_ppmm(p1, p2, a, b);
         p1 %= d;
         udiv_qrnnd(q, r2, p1, p2, d);
 
-        result = (r1 == r2);
+        int result = (r1 == r2);
```

Comment:
We *usually* declare variables as the good old days, i.e. in the beginning of the function. I know this differs somewhat (I know the `fft_small` module declares them where needed), but I think we should be consistent. Perhaps @fredrik-johansson could give his input when he comes back from his vacation.

Reply:
I'm indeed interested in viewpoints on this. As I wrote in another comment above, for these test files and some specific variables in them (loop counter, `result`, etc.) I totally understand that we want consistency and may want to declare them once for all at the beginning. But for core code (non-test, non-profile), I would not say the same, I find that in some places it makes the code harder to read and write because it makes less obvious --- sometimes far less when there many variables or long functions --- the scope of variables, whether they already have a value at some point, or even simply what their type is. Also in some pieces of code it even forbids us to do some things, like declaring some variables `const`. This being said, I guess (well, I hope) that the compiler is good enough for this to make no difference, so if there is a strong standpoint on the "old C style" declaration, I will follow it; in this case this should probably be written somewhere (I don't see it in the current code conventions in contributing.html).

Reply:
In the aspect of the compiler, I don't think it makes a difference. For me, it is about being able to study how many registers will be used, sort of. However, this obviously goes against the thought of a higher-level language. Moreover, I suspect I am one of the few that still prefers this way of declaring variables. And yes, whatever way we choose to go, it would be good if it is documented.